crates/sui-core/src/quorum_driver/reconfig_observer.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use async_trait::async_trait;
+use std::sync::Arc;
+use sui_types::committee::Committee;
+use tracing::{info, warn};
+
+use crate::{
+    authority::AuthorityStore,
+    authority_aggregator::{AuthAggMetrics, AuthorityAggregator},
+    authority_client::NetworkAuthorityClient,
+    epoch::committee_store::CommitteeStore,
+    safe_client::SafeClientMetricsBase,
+};
+
+use super::QuorumDriver;
+
+#[async_trait]
+pub trait ReconfigObserver<A> {
+    async fn run(&mut self, quorum_driver: Arc<QuorumDriver<A>>);
+}
+
+pub struct OnsiteReconfigObserver {
+    reconfig_rx: tokio::sync::broadcast::Receiver<Committee>,
+    authority_store: Arc<AuthorityStore>,
+    committee_store: Arc<CommitteeStore>,
+    safe_client_metrics_base: SafeClientMetricsBase,
+    auth_agg_metrics: AuthAggMetrics,
+}
+
+impl OnsiteReconfigObserver {
+    pub fn new(
+        reconfig_rx: tokio::sync::broadcast::Receiver<Committee>,
+        authority_store: Arc<AuthorityStore>,
+        committee_store: Arc<CommitteeStore>,
+        safe_client_metrics_base: SafeClientMetricsBase,
+        auth_agg_metrics: AuthAggMetrics,
+    ) -> Self {
+        Self {
+            reconfig_rx,
+            authority_store,
+            committee_store,
+            safe_client_metrics_base,
+            auth_agg_metrics,
+        }
+    }
+
+    async fn create_authority_aggregator_from_system_state(
+        &self,
+    ) -> AuthorityAggregator<NetworkAuthorityClient> {
+        AuthorityAggregator::new_from_system_state(
+            &self.authority_store,
+            &self.committee_store,
+            self.safe_client_metrics_base.clone(),
+            self.auth_agg_metrics.clone(),
+        )
+        // TODO: we should tolerate when <= f validators give invalid addresses
+        // GH issue: https://github.com/MystenLabs/sui/issues/7019
+        .unwrap_or_else(|e| {
+            warn!("panicked? {:?}", e);
+            panic!(
+                "Failed to create AuthorityAggregator from System State: {:?}",
+                e
+            )
+        })
+    }
+}
+
+#[async_trait]
+impl ReconfigObserver<NetworkAuthorityClient> for OnsiteReconfigObserver {
+    async fn run(&mut self, quorum_driver: Arc<QuorumDriver<NetworkAuthorityClient>>) {
+        // A tiny optimization: when the a very stale node just starts, the
+        // channel may fill up committees quickly. Here we skip directly to
+        // the last known committee by looking at SuiSystemState.
```

Comment:
As mentioned above, this seems to suggest that we can allow reconfig to proceed without us. Not sure I'm following how this is safe

Reply:
The optimization only optimizes the part that an `AuthorityAggregator` is created. For example. if a node is falling behind 100 epochs and then catching up, we do not need to create all 100 authority aggregators but only the latest one. What worries you in terms of safety?